### PR TITLE
daemon: Do not require native routing CIDR if ipmasq-agent is enabled

### PR DESCRIPTION
**Pull request by gandro**
When running with ipmasq-agent-based eBPF masquerading, there should be no need to mandatory require a native routing CIDR, since all CIDRs can be configured via the BPF map instead.

This PR therefore does not set the `IPV4_SNAT_EXCLUSION_DST_CIDR` field if no native routing CIDR is set. Notably, it also does not fall back on the local pod CIDR like we do in regular BPF masquearding, since in many IPAM modes there is no local pod CIDR to begin with.

Note that this change is backwards-compatible, because we have a check that ensures that the native routing CIDR is always set if BPF masquerading is enabled. So all existing users of ip-masq-agent already have a native routing CIDR set. This PR also removes that check, thus allowing future uses to not set a native routing CIDR when running with ipmasq-agent based masquerading.

### CODE REVIEW

**Review by christarazi — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by gandro** [comment]
There seems to have been some infra related flakes:

- ci-ipsec-upgrade https://github.com/cilium/cilium/actions/runs/6002359913/job/16278558972 failed downloading the helm charts
- ci-clustermesh https://github.com/cilium/cilium/actions/runs/6002358831/job/16278555174 failed reaching 1.1.1.1,
- ci-e2e https://github.com/cilium/cilium/actions/runs/6002358836/job/16278553484 failed reaching 1.1.1.1
- ci-ipsec-e2e https://github.com/cilium/cilium/actions/runs/6002359028/job/16278554830  failed reaching one.one.one.one
- runtime https://github.com/cilium/cilium/actions/runs/6002359548/job/16278594058 hit https://github.com/cilium/cilium/issues/26669

File: pkg/option/config.go [hunk 1]
```diff
@@ -3772,26 +3772,36 @@ func (c *DaemonConfig) checkMapSizeLimits() error {
...
-			IPAM, c.IPAMMode(), EnableIPv4Name)
+				"in combination with --%s=true --%s=true --%s=false --%s=%s --%s=%s",
+			IPv4NativeRoutingCIDR,
+			EnableIPv4Name, EnableIPv4Masquerade,
+			EnableIPMasqAgent,
+			RoutingMode, RoutingModeNative,
+			IPAM, c.IPAMMode())
 	}
 
 	return nil
```
Comment (pippolo84) [inline pkg/option/config.go]: Given that this helper is now pretty complicated, I'm wondering if reversing the conditions and using early return might improve readability:

```
	if c.GetIPv4NativeRoutingCIDR() != nil {
		return nil
	}
	if !c.EnableIPv4 || !c.EnableIPv4Masquerade {
		return nil
	}
	if c.EnableIPMasqAgent {
		return nil
	}
	if c.TunnelingEnabled() {
		return nil
	}
	if c.IPAMMode() == ipamOption.IPAMENI || c.IPAMMode() == ipamOption.IPAMAlibabaCloud {
		return nil
	}

	return fmt.Errorf(
		"native routing cidr must be configured with option --%s "+
			"in combination with --%s=true --%s=true --%s=false --%s=%s --%s=%s",
		IPv4NativeRoutingCIDR,
		EnableIPv4Name, EnableIPv4Masquerade,
		EnableIPMasqAgent,
		RoutingMode, RoutingModeNative,
		IPAM, c.IPAMMode())
```

File: pkg/option/config.go [hunk 2]
```diff
@@ -3772,26 +3772,36 @@ func (c *DaemonConfig) checkMapSizeLimits() error {
...
-			IPv6NativeRoutingCIDR, EnableIPv6Masquerade, RoutingMode, RoutingModeNative,
-			EnableIPv6Name)
+				"in combination with --%s=true --%s=true --%s=false --%s=%s",
+			IPv6NativeRoutingCIDR,
+			EnableIPv6Name, EnableIPv6Masquerade,
+			EnableIPMasqAgent,
+			RoutingMode, RoutingModeNative)
 	}
 
 	return nil
```
Comment (pippolo84) [inline pkg/option/config.go]: Same here with:
```
	if c.GetIPv6NativeRoutingCIDR() != nil {
		return nil
	}
	if !c.EnableIPv6 || !c.EnableIPv6Masquerade {
		return nil
	}
	if c.EnableIPMasqAgent {
		return nil
	}
	if c.TunnelingEnabled() {
		return nil
	}
	return fmt.Errorf(
		"native routing cidr must be configured with option --%s "+
			"in combination with --%s=true --%s=true --%s=false --%s=%s",
		IPv6NativeRoutingCIDR,
		EnableIPv6Name, EnableIPv6Masquerade,
		EnableIPMasqAgent,
		RoutingMode, RoutingModeNative)
```

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks!
I've left a non-blocking suggestion inline, feel free to ignore that if you think the current version is more readable.

File: pkg/option/config.go (hunk 1 above)
Comment (gandro) [reply pkg/option/config.go]: Yeah good point! I'll change it

**Review by qmonnet — APPROVED** [review]

**Review by ti-mo — APPROVED** [review]

**Review by derailed — APPROVED** [review]
@gandro LGTM Nice work Sebastian! 
Note: seeing a lot of parallels and similar logic in our codebase related to ipv4/6 handling. Wondering if we could apply generics for these use cases to dry up our code??

**Comment by gandro** [comment]
> @gandro LGTM Nice work Sebastian! Note: seeing a lot of parallels and similar logic in our codebase related to ipv4/6 handling. Wondering if we could apply generics for these use cases to dry up our code??

Good point. I think there is some experimentation needed first though

**Pull request merged**